md5sum_test.cpp

```diff
@@ -0,0 +1,326 @@
+        );
+
+    BOOST_TEST_EQUAL
+        (md5_calculate_file_checksum(md5sums_filename, md5_file_mode::text)
+        ,md5_calculate_file_checksum(md5sums_filename, md5_file_mode::binary)
+        );
+
+    BOOST_TEST_THROW
+        (md5_calculate_file_checksum("_ghost_")
+        ,std::runtime_error
+        ,"'_ghost_': no such file or directory"
+        );
+}
+
+/// Test md5_calculate_stream_checksum and md5_calculate_file_checksum
+/// functions.
+
+void MD5SumTest::TestMD5Reading() const
+{
+    std::vector<md5sum_for_file> md5sums
+        {{"md5_test", "e9f0fbd7a758d253ecd48ccded29234c", md5_file_mode::text}
+        ,{"md5_test", "e9f0fbd7a758d253ecd48ccded29234c", md5_file_mode::binary}
+        };
+
+    // Test md5_read_checksum_stream function.
+    std::istringstream is_sums{md5sums_text};
+    BOOST_TEST_EQUAL
+        (md5sums
+        ,md5_read_checksum_stream(is_sums, md5sums_filename)
+        );
+
+    std::istringstream is_throw1
+        {std::string{"e9f0fbd7a758d253ecd48ccded29234c  \n"}};
+    BOOST_TEST_THROW
+        (md5_read_checksum_stream(is_throw1, "test1")
+        ,std::runtime_error
+        ,"'test1': line too short at line 1"
+        );
+
+    std::istringstream is_throw2
+        {std::string{"efb7a0a972b88bb5b9ac6f60390d61bf_test\n"}};
+    BOOST_TEST_THROW
+        (md5_read_checksum_stream(is_throw2, "test2")
+        ,std::runtime_error
+        ,"'test2': incorrect checksum line format at line 1"
+        );
+
+    std::istringstream is_throw3
+        {std::string{"efb7a0a972b88bb5b9ac6f60390d61bf test\n"}};
+    BOOST_TEST_THROW
+        (md5_read_checksum_stream(is_throw3, "test3")
+        ,std::runtime_error
+        ,"'test3': incorrect checksum line format at line 1"
+        );
+
+    std::istringstream is_throw4
+        {std::string{"efb7a0a972b88bb5b9ac6f60390d61b  test\n"}};
+    BOOST_TEST_THROW
+        (md5_read_checksum_stream(is_throw4, "test4")
+        ,std::runtime_error
+        ,"'test4': incorrect MD5 sum format at line 1"
+        );
+
+    // Test md5_read_checksum_file function.
+    BOOST_TEST_EQUAL(md5sums, md5_read_checksum_file(md5sums_filename));
+
+    BOOST_TEST_THROW
+        (md5_read_checksum_file("_ghost_")
+        ,std::runtime_error
+        ,"'_ghost_': no such file or directory"
+        );
+}
+
+/// Test md5_hex_string function.
+
+void MD5SumTest::TestMD5ToHexString() const
+{
+    std::vector<unsigned char> v
+        {0x00, 0x10, 0x20, 0x30, 0x40, 0x50, 0x60, 0x70
+        ,0x80, 0x90, 0xa0, 0xb0, 0xc0, 0xd0, 0xe0, 0xf0
+        };
+    BOOST_TEST_EQUAL("00102030405060708090a0b0c0d0e0f0", md5_hex_string(v));
+}
+
+void MD5SumTest::RemoveTestFiles(char const* file, int line) const
+{
+    std::vector<std::string> filenames;
+    filenames.push_back(test_filename);
+    filenames.push_back(md5sums_filename);
+    for(auto const& i : filenames)
+        {
+        std::remove(i.c_str());
```

Comment:
It's ok if file already missing when we try to delete it (so there is no reason to check `remove()`s return value). I can simple get rid of `fs::exists()` checks.
```cpp
/// Before writing any test file, remove any old copy that may be left
/// over from a previous run that failed to complete, because old
/// copies can cause spurious error reports.

MD5SumTest::MD5SumTest()
{
    RemoveTestFiles(__FILE__, __LINE__);
```

Reply:
Maybe it would be better to call this function `RemoveTestFilesIfNecessary()` then. And the best would arguably be to check for existence and only call `remove()` if it does exist -- and then also check for its error.